org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/contentassist/ChainCompletionProposalComputer.java

```diff
@@ -0,0 +1,326 @@
+		final ChainFinder finder = new ChainFinder(expectedTypes, Arrays.asList(excludedTypes), invocationType);
+		final ExecutorService executor = Executors.newSingleThreadExecutor();
+		try {
+			Future<?> future = executor.submit(() -> {
+				if (findEntrypoints(expectedTypes, cu.getJavaProject())) {
+					finder.startChainSearch(entrypoints, maxChains, minDepth, maxDepth);
+				}
+			});
+			long timeout = Long.parseLong(JavaManipulation.getPreference("recommenders.chain.timeout", cu.getJavaProject()));
+			future.get(timeout, TimeUnit.SECONDS);
+		} catch (final Exception e) {
+			finder.cancel();
+			executor.shutdownNow();
+		}
+
+		return buildCompletionProposals(finder.getChains());
+	}
+
+	private List<CompletionItem> buildCompletionProposals(final List<Chain> chains) {
+		final List<CompletionItem> proposals = new LinkedList<>();
+
+		for (final Chain chain : chains) {
+			proposals.add(create(chain));
+		}
+		return proposals;
+	}
+
+	private boolean findEntrypoints(List<ChainType> expectedTypes, IJavaProject project) {
+		entrypoints = new LinkedList<>();
+		for (CompletionProposal prop : coll.getProposals()) {
+			IJavaElement javaElement = resolveJavaElement(prop, cu.getJavaProject());
+			if (javaElement != null) {
+				IJavaElement e = javaElement;
+				if (matchesExpectedPrefix(e) && !ChainFinder.isFromExcludedType(Arrays.asList(excludedTypes), e)) {
+					ChainElement ce = new ChainElement(e, false);
+					if (ce.getElementType() != null) {
+						entrypoints.add(ce);
+					}
+				}
+			} else {
+				IJavaElement[] visibleElements = coll.getContext().getVisibleElements(null);
+				for (IJavaElement ve : visibleElements) {
+					if (matchesExpectedPrefix(ve) && !ChainFinder.isFromExcludedType(Arrays.asList(excludedTypes), ve)) {
+						ChainElement ce = new ChainElement(ve, false);
+						if (ce.getElementType() != null) {
+							entrypoints.add(ce);
+						}
+					}
+				}
+			}
+		}
+		try {
+			entrypoints.addAll(computeContextEntrypoints(expectedTypes, project));
+		} catch (JavaModelException e) {
+			// continue
+		}
+
+		return !entrypoints.isEmpty();
+	}
+
+	private IJavaElement resolveJavaElement(CompletionProposal prop, IJavaProject proj) {
+		try {
+			if (prop.getKind() == CompletionProposal.FIELD_REF) {
+				return JDTUtils.resolveField(prop, proj);
+			} else if (prop.getKind() == CompletionProposal.METHOD_REF || prop.getKind() == CompletionProposal.ANNOTATION_ATTRIBUTE_REF) {
+				return JDTUtils.resolveMethod(prop, proj);
+			}
+		} catch (JavaModelException e) {
+			// continue
+		}
+		return null;
+	}
+
+	private boolean shouldPerformCompletionOnExpectedType() {
+		CompilationUnit cuNode = getASTRoot();
+		AST ast = cuNode.getAST();
+		return ast.resolveWellKnownType(ChainElementAnalyzer.getExpectedFullyQualifiedTypeName(coll.getContext())) != null || ChainElementAnalyzer.getExpectedType(cu.getJavaProject(), coll.getContext()) != null;
```

Comment:
Also, I wonder if we should opt to avoid doing the completion on all primitives, `String` ? I just don't want to be returning values that are useless.

Reply:
Yes lets stop primitives, String and Boxed Types.